### ISSUE TITLE
[clarification] Baseline C/C++-standard for code contributions

### ISSUE THREAD

**Issue by std-any-emplace** [zeromq/libzmq#3791]
If one were to make code-contributions, what is the accepted C or C++-standard? Are there any other compiler-/platform-specifics someone should be aware of? This information seems to be missing in https://github.com/zeromq/libzmq/blob/master/.github/CONTRIBUTING.md and http://wiki.zeromq.org/docs:contributing.

**Comment by sigiesec** [comment]
The C++ standard version is C++03.

There certainly are some additional aspects that arise from the support for not fully compliant compilers such as older Sun compilers, but the experiences with that are not documented.

**Comment by bluca** [comment]
> The C++ standard version is C++03.

I think we are supporting 98 too?

**Comment by sigiesec** [comment]
I am not sure. Do we have any platform that doesn't even support C++03? Admittedly I don't really know what the differences are exactly.

**Comment by std-any-emplace** [comment]
IMHO the most significant difference is that you have std::tr1::* available (completeness may depend on compiler). Among others you get tr1::shared_ptr, tr1::random, tr1::regex and some others (see for example https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_headers.html)

**Comment by sigiesec** [comment]
I don't think that's correct. TR1 is not part of C++03. Its first draft in from 2005 AFAIK. You can't use that unconditionally in libzmq. Note that there are some parts that conditionally depend on C++11 (atomics, e.g.). 

Does your original question refer to some specific topic/part of the standard? (Just curious, I completely agree that this should be specified more clearly.)

**Comment by std-any-emplace** [comment]
I was asking myself this question because looking at the code, there are quite some things in the core, where it is not obvious why things where done the way they were done or how they are working. Either its because, there are no/few comments or (and that's where my question is coming from) it uses some code which roughly duplicates functionality which is part of later c++-standards. And that duplicated functionality makes some parts not easy to understand because it differs from what I may know and assume from the libstd++-C++11-implementation for example for std::mutex or std::atomic.

In the end, looking at the comments here, that is a necessity for supporting platforms down to c++98/03.

**Comment by bluca** [comment]
> I am not sure. Do we have any platform that doesn't even support C++03? Admittedly I don't really know what the differences are exactly.

There are some old embedded things and old unixes which only have old gcc versions, with only C++ 98 - in fact some time ago a CI job that tests 98 compat was added:

https://github.com/zeromq/libzmq/blob/master/.travis.yml#L124

**Comment by sigiesec** [comment]
@std-any-emplace Yes, being able to use C++11/14/17 would make some things significantly easier, but as Luca wrote, we need to support some platforms with very old toolchains. https://github.com/zeromq/libzmq/blob/master/README.md attempts to list the supported platforms, but that list is also far from being perfect.

@bluca Ah, thanks for reminding me of that. In fact, now you mention it I remember having it seen before.

**Comment by std-any-emplace** [comment]
Despite the aforementioned recent addition of the test for c++98, is that necessity backed by evidence that there is in fact a user-base with antique toolchains using libzmq?

**Comment by bill-torpey** [comment]
On a related note, we're attempting to force a particular gnu99/gnu++98 standards (given that our own legacy code base is currently stuck on that):

```
CMAKE_C_FLAGS="-fno-omit-frame-pointer -std=gnu99"
CMAKE_CXX_FLAGS="-fno-omit-frame-pointer -std=gnu++98 -D_GLIBCXX_USE_CXX11_ABI=0 -Werror=reorder"
... 
cmake -DCMAKE_CXX_FLAGS="${CMAKE_CXX_FLAGS}" -DCMAKE_C_FLAGS="${CMAKE_C_FLAGS}" ...

```

But it doesn't work -- since CMakeLists.txt sets it specifically:

```
check_cxx_compiler_flag("-std=gnu++11" COMPILER_SUPPORTS_CXX11)
if(COMPILER_SUPPORTS_CXX11)
  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++11")
endif()
check_c_compiler_flag("-std=gnu11" COMPILER_SUPPORTS_C11)
if(COMPILER_SUPPORTS_C11)
  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu11")
else()
  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -std=gnu99")
endif()
```
 We end up with both settings on the command line, but the compiler honors the last (wrong) one:

```
/build/share/clang/10.x/bin/clang++   -DZMQ_BUILD_DRAFT_API -DZMQ_CUSTOM_PLATFORM_HPP -DZMQ_HAVE_ATOMIC_INTRINSICS -DZMQ_HAVE_TIPC -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -fno-omit-frame-pointer -std=gnu++98 -D_GLIBCXX_USE_CXX11_ABI=0 -Werror=reorder  -fno-omit-frame-pointer -fsanitize=undefined -fsanitize=implicit-conversion -fsanitize=implicit-integer-truncation -fsanitize=integer -fsanitize=nullability -fsanitize=vptr -std=gnu++11 -Wno-tautological-constant-compare -Wall -Wextra -Wno-long-long -Wno-uninitialized -pedantic -O2 -g -DNDEBUG -fPIC -I/home/btorpey/work/libzmq/master/src/include -I/home/btorpey/work/libzmq/master/ubsan    -o CMakeFiles/objects.dir/src/session_base.cpp.o -c /home/btorpey/work/libzmq/master/src/src/session_base.cpp
```

So is it safe to say that forcing C++11 semantics, just because the compiler supports it, is a bug?

**Comment by bluca** [comment]
If you want to switch the parameters around, so that the user-supplied ones come last, that would be fine

**Comment by bill-torpey** [comment]
Leaving aside how to do that (I guess scanning the CMAKE_CXX_FLAGS for "-std=" to see if it's already been set), the question remains.

libzmq is forcing C++11 semantics if the compiler supports it.  Is that intended behavior?

**Comment by bluca** [comment]
Yes - there's a chunk of backward-compatible local reimplementations or slower alternatives to what more modern standard provides, so if the compiler supports it it's good to use it

**Comment by bill-torpey** [comment]
OK -- how about this: https://github.com/zeromq/libzmq/pull/3909

**Comment by bluca** [comment]
Fixed by #3909
